internal/controllers/utils/events.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"sort"
+	"time"
+)
+
+const (
+	// Generic Event states
+	// InProgress indicates a in progress state
+	InProgress = "InProgress"
+	// InProgress indicates a completed state
+	Completed = "Completed"
+)
+
+// Event is the generic representation of an event
+type Event struct {
+	ObjectID  string
+	Timestamp time.Time
+	State     string
+}
+
+// EventHistory contains a slice of events
+type EventHistory struct {
+	History []*Event
+}
+
+// IsTimedOut proceses a process history an determines the overall state implied
+// by the events.
+// returns true if for any ObjectID:
+// - the overall state was InProgress from the initial time for more than timeout
+// - the overall state was InProgress from the last transition to InProgress for more than timeout
+// returns false if for any ObjectID:
+// - the last transition is Completed
+// - no events to process
+func (h EventHistory) IsTimedOut(now time.Time, timeout time.Duration) bool {
+	// Sort events chronologically from old to new
+	sort.Slice(h.History, func(i, j int) bool {
+		return h.History[i].Timestamp.Before(h.History[j].Timestamp)
+	})
+
+	// currentState stores the running value of the State (InProgress or Completed) for each ObjectID
+	currentState := map[string]string{}
+	// Initialize the current state to NonCompliant
+	for _, event := range h.History {
+		currentState[event.ObjectID] = InProgress
```

Comment:
The comment above says `NonCompliant`, but the code sets it to `InProgress`.

Reply:
Fixed